packager/file/http_file_unittest.cc

```diff
@@ -64,176 +66,230 @@ nlohmann::json HandleResponse(const FilePtr& file) {
   return value;
 }
 
+// Tests using httpbin can sometimes be flaky.  We get HTTP 502 errors when it
+// is overloaded.  This will retry a test with delays, up to a limit, if the
+// HTTP status code is 502.
+void RetryTest(std::function<HttpFile*()> setup,
+               std::function<void(FilePtr&)> pre_read,
+               std::function<void(FilePtr&, nlohmann::json)> post_read) {
+  nlohmann::json json;
+  FilePtr file;
+
+  for (int i = 0; i < 3; ++i) {
+    file.reset(setup());
+
+    ASSERT_TRUE(file->Open());
+
+    pre_read(file);
+
+    json = HandleResponse(file);
+
+    if (file->HttpStatusCode() != 502) {
+      // Not a 502 error, so take this result.
+      break;
+    }
+
+    // Delay with exponential increase (1s, 2s, 4s), then loop try again.
+    int delay = 1 << i;
+    LOG(WARNING) << "httpbin failure (" << file->HttpStatusCode() << "): "
+                 << "Delaying " << delay << " seconds and retrying.";
+    std::this_thread::sleep_for(std::chrono::seconds(delay));
+  }
+
+  // Out of retries?  Check what we have.
+  post_read(file, json);
+}
+
 }  // namespace
 
 TEST(HttpFileTest, BasicGet) {
-  FilePtr file(new HttpFile(HttpMethod::kGet, "https://httpbin.org/anything"));
-  ASSERT_TRUE(file);
-  ASSERT_TRUE(file->Open());
-
-  auto json = HandleResponse(file);
-  ASSERT_TRUE(json.is_object());
-  ASSERT_TRUE(file.release()->Close());
-  ASSERT_JSON_STRING(json, "method", "GET");
+  RetryTest(
+      []() -> HttpFile* {
+        return new HttpFile(HttpMethod::kGet, "https://httpbin.org/anything");
+      },
+      [](FilePtr&) -> void {},
+      [](FilePtr& file, nlohmann::json json) -> void {
+        ASSERT_TRUE(json.is_object());
+        ASSERT_TRUE(file.release()->Close());
+        ASSERT_JSON_STRING(json, "method", "GET");
```

Comment:
nit: a slightly cleaner approach here and below: line 112 and 113 can be moved into RetryTest implementation, and line 114 can be checked after RetryTest. This eliminates the post_read() function callback.

Reply:
We can't move 112 into RetryTest, because the error tests will produce is_object() == false.

If we're making changes to the test structure, though, I would go farther with 113 and eliminate it.  If we don't care about testing Close() return value, it's not providing value.  FilePtr being a unique_ptr with a closer will take care of closing the HttpFile object no matter what.  I'll drop it completely.  If you feel we need to test Close() returning true, I'll add an explicit test case for that.  But it doesn't seem necessary to me at a glance.

Checking 114 after RetryTest would require us to return the json object and check results after the retry callback.  That only works if we don't need to do ASSERT_* in RetryTest, though.  ASSERT_* macros can only be used in void functions, since they return early on failure.  I currently have ASSERT_TRUE(file->Open()) in RetryTest, so I can't have json be a return value.

Also, some test cases want to verify things on file (error cases), while some want to verify things about json (success cases).  So I prefer the callback that provides both, while letting the RetryTest function manage the lifetime of the HttpFile object.

Reply:
I dropped the Close() check, but the other changes you suggested didn't work

Reply:
sg.